Features/task 1

RadekP3 opened the pull request.

RadekP3 committed:
KA-343 create Item model

RadekP3 committed:
KA-343 create dummy RESTful controller

RadekP3 committed:
KA-343 create item comparer for testing purposes

RadekP3 committed:
KA-343 add unit tests for ItemController

RadekP3 committed:
KA-343 use Guid instead of int for Ids

RadekP3 committed:
KA-343 append 'Async' to methods names

RadekP3 committed:
KA-343 implement URI versioning

RadekP3 committed:
KA-343 refactor tests using extension method

RadekP3 committed:
KA-343 override ToString in Item

RadekP3 committed:
KA-343 remove unused imports

RadekP3 committed:
KA-343 fix wrong namespace

RadekP3 committed:
KA-343 use Guid instead of int

PetrSvirak commented on TodoListApp/TodoList.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,50 @@
+        {
+            new Item() {Id = Guid.Empty, Text = "First"},
+            new Item() {Id = Guid.Empty, Text = "Second"},
+            new Item() {Id = Guid.Empty, Text = "Third"}
```
There are some unnecessary braces – embrace your R# 😉

PetrSvirak replied:
You cannot use `Guid.Empty` here, since they don't form a valid response – if your API went life and returned an array of such items, any user would probably ask themselves whether it really works – you need to provide static yet **correct** items…

PetrSvirak replied:
JSON is largely consumed by JavaScript applications and even your (only) consumer application is written in JS; JS's naming convention uses cammelCase while C#'s uses PascalCase when refering to object properties. While you must keep both conventions sound, your API should return JSON in format that would aplease most JS clients (as a courtesy)…

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,50 @@
+        };
+
+        // GET: api/v{version}/items
+        [Route("")]
```
You don't need this `Route` attribute since you have the same route defined at the `Controller` level…

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,50 @@
+
+
+        // GET: api/v{version}/items/5
+        [Route("{id:int}")]
```
This end-point will never get hit since the route looks for `int` but you decode a `Guid`… Have you manually tested your end-points?

RadekP3 replied:
I did test them in Postman before I changed int to Guid. I forgot to test it afterward, sorry about that.

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api/Models/Item.cs:
```diff
@@ -0,0 +1,14 @@
+        public string Text { get; set; } = "";
+        public override string ToString()
+        {
+            return $"Id: {Id}, Text: {Text}";
```
Return is the first statement `=>` arrow function 😉

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/Comparers/ItemComparer.cs:
```diff
@@ -0,0 +1,25 @@
+
+        public int GetHashCode(Item obj)
+        {
+            unchecked
```
Why is the `unchecked` here? (it really is the question → I don't urge you to remove the block, I urge you to answer the question 😃)

RadekP3 replied:
Allegedly, it improves performance, because overflow-checking for arithmetic operations is suppressed and therefore code execution is faster. And it's fine if the value overflows in this case.
But yeah, I only used it because R# suggested it to be honest.

RadekP3 replied:
Allegedly, code execution is faster when overflow-checking is suppressed. And it's fine if the value overflows in this case, because we don't really care how big or small the value of hashcode is, we just want it to be the same for the same objects.
But to be honest, I only used it because R# suggested it😃

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/Extensions/ControllerExtension.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public static class ControllerExtension
+    {
+        public static async Task<HttpResponseMessage> ExecuteAction(this ItemsController controller, Func<ItemsController, Task<IHttpActionResult>> action)
```
What if I had another controller, is there something specific that would prevent me from using the code in this very method (or is it just the type of the `controller` parameter)? If there really is a reason preventing other controllers from being passed then why is the file so generically named?

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+            new Item {Id = Guid.Empty, Text = "Second"},
+            new Item {Id = Guid.Empty, Text = "Third"}
+        };
+        readonly ItemComparer _comparer = new ItemComparer();
```
Your `ItemComparer` is a class that only contains pure functions and no state - kinda "pure class". This makes it an ideal candidate to be a singleton. While it is not technically an issue in this very case, you should prevent the `ItemComparer` from being unnecessary instantiated. Also, have a look at `Lazy` and utilize it - please, explain why you used it and what benefits it has in the commit message.

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+            new Item {Id = Guid.Empty, Text = "Third"}
+        };
+        readonly ItemComparer _comparer = new ItemComparer();
+        readonly Item _mockItem = new Item() { Id = Guid.Empty, Text = "Item" };
```
While it is a line-saving solution, it does not seem to necessarily describe true nature of your API - are you sure that `Put` and `Post` will always accept the same structure in their bodies?

PetrSvirak replied:
You don't anticipate a change to the item that can happen in a test - imagine if `controller.Post` called `item.Id = Guid.New();`: Then your `Put` test would be provided with different data…

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+        };
+        readonly ItemComparer _comparer = new ItemComparer();
+        readonly Item _mockItem = new Item() { Id = Guid.Empty, Text = "Item" };
+        readonly ItemsController _controller = new ItemsController() { Configuration = new HttpConfiguration(), Request = new HttpRequestMessage()};
```
First of all, this code should not all be squeezed on a single line.

PetrSvirak replied:
Second, you don't anticipate a change in the controller's state.

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+        public async Task GetAsync_ReturnsDefaultItemsWithOkStatusCode()
+        {
+            var actionResult = await _controller.ExecuteAction(controller => controller.GetAsync());
+            actionResult.TryGetContentValue(out Item[] itemsFromMessage);
```
Don't limit your test to arrays, if you change the type of returned value (e.g. you would need to call "ToList" for a reason in your controller), this test would fail - tests should not push production code into implementation details.

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+            actionResult.TryGetContentValue(out Item[] itemsFromMessage);
+
+            Assert.That(actionResult.StatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(itemsFromMessage, Is.EqualTo(_defaultItems).Using(_comparer));
```
That's OK you use single `_comparer` instance throughout the tests, however, it could be made even better if it was an extension method (in connection with the comment on [L23](https://github.com/RadekP3/kentico-onboarding-cs/pull/3/files#diff-d99e05ec2d814d50a21976a19c5a6fdeR23)) – as a bonus, it would allow you to reuse the comparer in other tests when the time comes.

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+            actionResult.TryGetContentValue(out Item itemFromMessage);
+
+            Assert.That(actionResult.StatusCode, Is.EqualTo(HttpStatusCode.Created));
+            Assert.That(itemFromMessage, Is.EqualTo(_defaultItems[0]).Using(_comparer));
```
You do not check the `createdAt` URI…

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+            Assert.That(actionResult.StatusCode, Is.EqualTo(HttpStatusCode.Created));
+            Assert.That(itemFromMessage, Is.EqualTo(_defaultItems[0]).Using(_comparer));
+        }
+        [Test]
```
An empty line missing between tests.

RadekP3 replied:
👍

PetrSvirak commented on TodoListApp/TodoList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -0,0 +1,75 @@
+            Assert.That(itemsFromMessage, Is.EqualTo(_defaultItems).Using(_comparer));
+        }
+        [Test]
+        public async Task GetAsync_Id_ReturnsFirstItemWithOkStatusCode()
```
You have to describe what kind of `Id` is actually passed to the item - how you qualify it or describe what it represents to a colleague - pick a suitable adjective or two.

RadekP3 replied:
👍

PetrSvirak reviewed (changes requested):
Your commits are reasonably named and there is a quite small amount of them. Cool 👍 

Feature-wise, I see some room for improvement; especially in the sense of response-correctness and reliability. Please, attend to the related issues below.

The code is mostly OK, yet you will encounter some comments that require further code-tinkering from your side.

One other thing I miss here is a R# configuration that would allow other participating developers to share the same code-style and practices. Please, add it to both your repository and solution (as a "solution file)…

RadekP3 committed:
KA-343 use proper Guid instead of just Guid.Empty

RadekP3 committed:
KA-343 make API return JSON

RadekP3 committed:
KA-343 fix mistakes in controller

RadekP3 committed:
KA-343 convert ToString to arrow function

RadekP3 committed:
KA-343 give extension method more suitable name

RadekP3 committed:
KA-343 make ItemComparer a singleton

Use Lazy<T> because it ensures that the instance of the ItemComparer
will only be created when it's actually needed.
The benefits or Lazy are:
 - it improves performance,
 - it avoids wasteful computation
 - it reduces program memory requirements

RadekP3 committed:
KA-343 use item comparer as extension method

RadekP3 committed:
KA-343 replace mockItem attribute with local var

Reasons:
 - Put and Post might want to accept different structures
 - Item could be changed in a test and e.g. Put and Post could
  than be provided with different data

RadekP3 committed:
KA-343 move controller initialization to SetUp

RadekP3 committed:
KA-343 reformat code and state accessability explicitly

RadekP3 committed:
KA-343 replace Array with IEnumerable

RadekP3 committed:
KA-343 check for created URI in Post test

RadekP3 committed:
KA-343 make tests more AAA

RadekP3 committed:
KA-343 desribe tests better

RadekP3 committed:
KA-343 add R# configuration file

RadekP3 committed:
KA-343 reformat code
